### DISCUSSION TITLE
Building libMoltenVK fails when linking with libMoltenVKShaderConverter.a

### DISCUSSION THREAD

**Discussion by gzorin** [KhronosGroup/MoltenVK#1405] [Q&A]
I'm trying to build MoltenVK for macOS. Building "libMoltenVK.a" fails for me, because "libMoltenVKShaderConverter.a" can't be found during the link phase; it's looking in "MoltenVK/MoltenVK/build/Release/libMoltenVKShaderConverter.a", but that library was instead written to various locations under "MoltenVK/MoltenVKShaderConverter/build".

This happens with the 'master' branch and various recent tags that I've tried, and whether I use 'xcodebuild' or try various build schemes within Xcode itself. I'm using Xcode 12.5.1 on macOS 11.4, & am interested in building the library for use on macOS.

Could someone please suggest what I'm doing wrong here?

**Comment by billhollings** [comment]
Those locations (`target/build` directory) look odd.

How are you trying to do the build?

The `README.md` document explains the Xcode and command line build options.

**Reply by gzorin** [reply]
Thank you for replying!

From Xcode itself, I've been opening "MoltenVKPackaging.xcodeproj", choosing the schemes "Molten VK Package (macOS only) > My Mac" and "MoltenVK-macOS > My Mac", followed by "Build" from the "Product" menu.

From the command line, from the root directory of the 'MoltenVK' repo checkout, I've tried stuff like `xcodebuild build -quiet -project MoltenVKPackaging.xcodeproj -scheme "MoltenVK Package (macOS only)" -configuration "Release"` and `xcodebuild build -quiet -project MoltenVKPackaging.xcodeproj -scheme "MoltenVK Package (macOS only)" -configuration "Debug"`, as well as `make macos` and `make macos-debug`.

In all cases, it fails because `ld` can't find `libMoltenVKShaderConverter.a`. I can see that `libMoltenVKShaderConverter.a` is being built, but in a different filesystem location.

**Reply by gzorin** [reply]
So since Homebrew makes version 1.1.4 available, I looked at their build scripts, & was able to avoid the linker error when building from the command line by passing values for `SYMROOT` and `OBJROOT`:

```
xcodebuild build -quiet -project MoltenVKPackaging.xcodeproj -scheme "MoltenVK Package (macOS only)" -configuration "Release" SYMROOT=/Users/alex/MoltenVK/build/ OBJROOT=build
```

**Comment by billhollings** [comment]
I'm glad you got it working in the end.

What happens when you run `make macos` from the command line within the root MoltenVK directory (see `README.md` for more on that)?

The MoltenVK build scripts should be agnostic to your Xcode build location settings. So it's surprising that you are encountering any issues at all. It looks like the default Xcode location settings build all binaries for a platform in the same directory, but your Xcode location settings seem to partition them by sub-project or target.

**Reply by gzorin** [reply]
`make macos` also fails with the same linker error.

**Comment by gzorin** [comment]
Looks like this was due to an Xcode preference (which I don't recall ever modifying). Under "Preferences > Locations", under "Advanced..." next to "Derived Data", I had "Legacy" chosen; after choosing "Unique" instead then the build processes as described in your README.
